File: solutions/1508. Range Sum of Sorted Subarray Sums/1508.cpp
// Time complexity: O(nlog(S))
// Space complexity: O(1)
class Solution {
 public:
  int rangeSum(const vector<int>& nums, int n, int left, int right) {
    constexpr int kMod = 1'000'000'007;

    auto subarraysAndSumNoGreaterThan = [&](int m) -> pair<int, long> {
      int count = 0;
      long total = 0;
      int sum = 0;
      int window = 0;

      for (int i = 0, j = 0; j < n; ++j) {
        sum += nums[j] * (j - i + 1);
        window += nums[j]; 
        while (window > m) {
          sum -= window;
          window -= nums[i++];
        }
        count += j - i + 1;
        total += sum;
      }

      return {count, total};
    };

    const int L = ranges::min(nums);
    const int R = accumulate(nums.begin(), nums.end(), 0);

    auto firstKSubarraysSum = [&](int k) -> long {
      int l = L;
      int r = R;

      while (l < r) {
        const int m = l + (r - l) / 2;
        if (subarraysAndSumNoGreaterThan(m).first < k)
          l = m + 1;
        else
          r = m;
      }

      const auto& [count, total] = subarraysAndSumNoGreaterThan(l);
      return total - l * (count - k);
    };

    return (firstKSubarraysSum(right) - firstKSubarraysSum(left - 1)) % kMod;
  }
};

File: solutions/1508. Range Sum of Sorted Subarray Sums/1508.rs
// Time complexity: O(nlog(S))
// Space complexity: O(1)
impl Solution {
    pub fn range_sum(nums: Vec<i32>, n: i32, left: i32, right: i32) -> i32 {
        const k_mod: i64 = 1_000_000_007;

        let subarrays_and_sum_no_greater_than = |nums: &Vec<i32>, m: i32| -> (i32, i64) {
            let mut count = 0;
            let mut total = 0;
            let mut sum = 0;
            let mut window = 0;
            let mut i = 0;

            for j in 0..nums.len() {
                sum += nums[j] * (j - i + 1) as i32;
                window += nums[j];

                while window > m {
                    sum -= window;
                    window -= nums[i];
                    i += 1;
                }

                count += (j - i + 1) as i32;
                total += sum as i64;
            }

            (count, total)
        };

        let l = *nums.iter().min().unwrap();
        let r = nums.iter().sum::<i32>();

        let first_k_subarrays_sum = |nums: &Vec<i32>, k: i32| -> i64 {
            let mut l = l;
            let mut r = r;

            while l < r {
                let m = l + (r - l) / 2;
                if subarrays_and_sum_no_greater_than(nums, m).0 < k {
                    l = m + 1;
                } else {
                    r = m;
                }
            }

            let (count, total) = subarrays_and_sum_no_greater_than(nums, l);
            total - l as i64 * (count - k) as i64
        };

        let left_sum = first_k_subarrays_sum(&nums, left - 1);
        let right_sum = first_k_subarrays_sum(&nums, right);

        ((right_sum - left_sum + k_mod) % k_mod) as i32
    }
}
